src/Features/Core/Portable/EditAndContinue/DebuggeeModuleMetadataCache.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.EditAndContinue
+{
+    /// <summary>
+    /// A cache of metadata blobs loaded into processes being debugged.
+    /// Thread safe.
+    /// </summary>
+    internal sealed class DebuggeeModuleMetadataCache
+    {
+        // Maps MVIDs to metadata blobs loaded to specific processes
+        private Dictionary<Guid, ModuleMetadata> _lazyCache;
+
+        /// <summary>
+        /// May return null if the provider returns null.
+        /// </summary>
+        public ModuleMetadata GetOrAdd(Guid mvid, Func<Guid, ModuleMetadata> provider)
+        {
+            if (_lazyCache == null)
+            {
+                Interlocked.CompareExchange(ref _lazyCache, new Dictionary<Guid, ModuleMetadata>(), null);
+            }
+
+            var cache = _lazyCache;
+
+            lock (cache)
+            {
+                if (cache.TryGetValue(mvid, out var existing))
+                {
+                    return existing;
+                }
+            }
+
+            var newMetadata = provider(mvid);
+            if (newMetadata == null)
+            {
+                return null;
+            }
```

Comment:
The module may be available next time we ask. Modules are loaded and unloaded to the debuggee arbitrarily during debugging session.

---
In reply to: [133052213](https://github.com/dotnet/roslyn/pull/21466#discussion_r133052213) [](ancestors = 133052213)